lib/auth0/mixins/validation.rb

```diff
@@ -20,6 +28,118 @@ def validate_permissions_array(permissions)
+          when Auth0::Algorithm::HS256
+            secret = algorithm.secret
+          end
+
+          begin
+            JWT.decode decoded_id_token[:string], secret, true, options
+          rescue JWT::VerificationError
+            raise Auth0::InvalidIdToken, 'Invalid ID token signature'
+          rescue JWT::IncorrectAlgorithm
+            alg = decoded_id_token[:header]['alg']
+            raise Auth0::InvalidIdToken, "Signature algorithm of \"#{alg}\" is not supported. Expected the ID token to be signed with \"#{algorithm_name}\""
+          rescue JWT::DecodeError
+            kid = decoded_id_token[:header]['kid']
+            raise Auth0::InvalidIdToken, "Could not find a public key for Key ID (kid) \"#{kid}\""
+          end
+        end
+
+        def validate_claims(decoded_id_token); end
+      end
+
+      module Algorithm
+        # Represents the HS256 algorithm, which rely on shared secrets.
+        # @see https://auth0.com/docs/tokens/concepts/signing-algorithms
+        class HS256
+          class << self
+            private :new
+
+            # Create a new instance passing the shared secret.
+            # @param secret [string] The HMAC shared secret.
+            # @return [object] A new instance.
+            def secret(secret)
+              new secret
+            end
+          end
+
+          attr_accessor :secret
+
+          def initialize(secret)
+            raise Auth0::InvalidParameter, 'Must supply a valid secret' if secret.to_s.empty?
+
+            @secret = secret
+          end
+        end
+
+        # Represents the RS256 algorithm, which rely on public key certificates.
+        # @see https://auth0.com/docs/tokens/concepts/signing-algorithms
+        class RS256
+          include Auth0::Mixins::HTTPProxy
+
+          @@cache = Zache.new.freeze
+
+          class << self
+            private :new
+
+            # Create a new instance passing the shared secret.
+            # @param secret [string] The url where the JWK set is located.
+            # @param lifetime [integer] The lifetime of the JWK set in-memory cache in seconds.
+            #   Must be a non-negative value. Defaults to 600 seconds (10 minutes).
+            # @return [object] A new instance.
+            def jwks_url(url, lifetime: 10 * 60)
+              new url, lifetime
+            end
+          end
+
+          def initialize(jwks_url, lifetime)
+            raise Auth0::InvalidParameter, 'Must supply a valid jwks_url' if jwks_url.to_s.empty?
+            raise Auth0::InvalidParameter, 'Must supply a valid lifetime' unless lifetime.is_a?(Integer) && lifetime >= 0
+
+            @lifetime = lifetime
+            @jwks_url = jwks_url
+          end
+
+          # Fetches the JWK set from the in-memory cache, or from the url as a fallback.
+          # @return [hash] A JWK set.
+          def jwks
+            previous_value = @@cache.last(:jwks)[:value] unless @@cache.last(:jwks).nil?
```

Comment:
what is this for, why do you need to invoke `last` on the cache? and what's the value of `:jwks` and `:value`?

Reply:
That was added as a logical safety measure to make sure I do not end up caching a `nil` value, in the event that the network request fails in a way that does not throw an exception (e.g an unexpected response with a successful status code, like sometimes happens in an outage). In that case, the cache will keep the last known value for another `lifetime` period.
I added the `last` method to the cache class to be able to just fetch the last cached value, expired or not.
`:jwks` is the cache key for the jwks. `:value` is the key to access the actual jwks content in the cache entry, that is a hash containing other [values](https://github.com/yegor256/zache/blob/master/lib/zache.rb#L193) (the start time and the lifetime).

Reply:
That's good thanks for clarifying. Remember the cache miss scenario I mentioned in the general review comment.

Reply:
Though `:value` should be in the definition of `last` though. I'll change it.

Reply:
Chaged in efdf3feec312ed9d2e8a91fa3ea9b625373033ec